src/client.rs

```diff
@@ -126,6 +137,45 @@ impl Client {
         Ok(response)
     }
 
+    /// Sends a batch of requests to the client.  The return vector holds the response
+    /// for the request at the corresponding index.  If no response was provided, it's [None].
+    ///
+    /// Note that the requests need to have valid IDs, so it is advised to create the requests
+    /// with [build_request].
+    pub fn send_batch(&self, requests: &[Request]) -> Result<Vec<Option<Response>>, Error> {
+        if requests.len() < 1 {
+            return Err(Error::EmptyBatch);
+        }
+
+        // If the request body is invalid JSON, the response is a single response object.
+        // We ignore this case since we are confident we are producing valid JSON.
+        let responses: Vec<Response> = self.send_raw(&requests)?;
+        if responses.len() > requests.len() {
+            return Err(Error::WrongBatchResponseSize);
+        }
+
+        // First index responses by ID and catch duplicate IDs.
+        let mut resp_by_id = HashMap::new();
+        for resp in responses.into_iter() {
+            // IDs can only be of JSON type String or Number (or Null),
+            // so converting to string should be inexpensive.
```

Comment:
FYI it's guaranteed to result in an allocation/deallocation pair, which could fragment memory and be very expensive indeed.

Reply:
Comment should be updated now to "cloning should be inexpensive and require no allocations".